mitmproxy/addons/proxyserver.py

```diff
@@ -0,0 +1,137 @@
+class ProxyConnectionHandler(server.StreamConnectionHandler):
+    master: master.Master
+
+    def __init__(self, master, r, w, options):
+        self.master = master
+        super().__init__(r, w, options)
+        self.log_prefix = f"{human.format_address(self.client.peername)}: "
+
+    async def handle_hook(self, hook: commands.Hook) -> None:
+        with self.timeout_watchdog.disarm():
+            # We currently only support single-argument hooks.
+            data, = hook.args()
+            data.reply = AsyncReply(data)
+            await self.master.addons.handle_lifecycle(hook.name, data)
+            await data.reply.done.wait()
+            data.reply = None
+
+    def log(self, message: str, level: str = "info") -> None:
+        x = log.LogEntry(self.log_prefix + message, level)
+        x.reply = controller.DummyReply()  # type: ignore
+        asyncio_utils.create_task(
+            self.master.addons.handle_lifecycle("log", x),
+            name="ProxyConnectionHandler.log"
+        )
+
+
+class Proxyserver:
+    """
+    This addon runs the actual proxy server.
+    """
+    server: Optional[asyncio.AbstractServer]
+    listen_port: int
+    master: master.Master
+    options: options.Options
+    is_running: bool
+
+    def __init__(self):
+        self._lock = asyncio.Lock()
+        self.server = None
+        self.is_running = False
+
+    def load(self, loader):
+        loader.add_option(
+            "connection_strategy", str, "lazy",
+            "Determine when server connections should be established.",
+            choices=("eager", "lazy")
+        )
+        loader.add_option(
+            "proxy_debug", bool, False,
+            "Enable debug logs in the proxy core.",
+        )
+        # FIXME: Update allowed events to include new ones.
+        eventsequence.Events = frozenset(
+            eventsequence.Events | set(commands.all_hooks.keys())
+        )
+
+    def running(self):
+        self.master = ctx.master
+        self.options = ctx.options
+        self.is_running = True
+        self.configure(["listen_port"])
+
+    def configure(self, updated):
+        if not self.is_running:
+            return
+        if "mode" in updated and ctx.options.mode == "transparent":  # pragma: no cover
+            platform.init_transparent_mode()
+        if any(x in updated for x in ["server", "listen_host", "listen_port"]):
+            asyncio.create_task(self.refresh_server())
+
+    async def refresh_server(self):
+        async with self._lock:
+            if self.server:
+                await self.shutdown_server()
+                self.server = None
+            if ctx.options.server:
+                self.server = await asyncio.start_server(
+                    self.handle_connection,
+                    self.options.listen_host,
+                    self.options.listen_port,
+                )
+                addrs = {f"http://{human.format_address(s.getsockname())}" for s in self.server.sockets}
```

Comment:
Do we already support the HTTPS secure proxy format?

Reply:
We now do! Let's do some testing how that works out. Practically speaking your client needs to trust the cert for this already, so that's a bit of a chicken-egg problem for mitm.it.